### PR TITLE
Add monitorInfo class using native capabilities to replace WinForms s…

### PR DESCRIPTION
**Pull request by Averus-a**
…creens API

### Issues Resolved ### 

- fixes #

### API Changes ###

 
None

### Platforms Affected ### 

- All
- WPF
- UWP
- iOS
- Android

### Behavioral Changes ###

None

### PR Checklist ###

- [x] I have included examples or tests
- [ ] I have updated the change log
- [ ] I am listed in the CONTRIBUTORS file
- [ ] Rebased on top of the target branch at time of PR
- [x] Changes adhere to coding standard

### CODE REVIEW

File: src/Orchestra.Core/Windows/MonitorInfo.cs [hunk 1]
```diff
@@ -0,0 +1,452 @@
...
+        }
+
+        public static MonitorInfo GetMonitorFromWindowHandle(IntPtr handle)
+        {
+            // Get screen from window handle
+            var monitorHandle = NativeMethods.MonitorFromWindow(handle, 0);
+
+            var nativeInfo = new NativeMonitorInfoEx();
+           
+            var success = NativeMethods.GetMonitorInfo(monitorHandle, nativeInfo);
```
Comment (GeertvanHorrik) [inline src/Orchestra.Core/Windows/MonitorInfo.cs:185]: Ref native Info?

File: src/Orchestra.Core/Windows/MonitorInfo.cs [hunk 2]
```diff
@@ -0,0 +1,452 @@
...
+
+            if(!success)
+            {
+                return null;
+            }
+
+            // note: can this cause issues when monitor mirror, probably need to check DisplayDeviceStateFlag?
+            var outputDevice = GetOutputDevicesForDevice(nativeInfo.GetDeviceName()).FirstOrDefault();
```
Comment (GeertvanHorrik) [inline src/Orchestra.Core/Windows/MonitorInfo.cs:193]: If we do First Or Default, we should check for null afterwards

**Comment by GeertvanHorrik** [comment]
Most is reviewed, only need to check the Wpf Screen on an actual pc first

File: src/Orchestra.Core/Windows/MonitorInfo.cs [hunk 3]
```diff
@@ -0,0 +1,452 @@
...
+            {
+                var mi = new NativeMonitorInfoEx();
+                var success = NativeMethods.GetMonitorInfo(hMonitor, mi);
+
+                if (success)
+                {
+                    nativeMonitorInfos.Add(mi);
+                }
+
+                return true;
```
[the remark this answers is not in the document]
Comment (Averus-a) [reply src/Orchestra.Core/Windows/MonitorInfo.cs:135]: Will change the behavior of function - enumeration stops on false value end you don't get results even if other monitor handles are correct.

File: src/Orchestra.Core/Windows/MonitorInfo.cs [hunk 4]
```diff
@@ -0,0 +1,452 @@
...
+
+        public bool IsPrimary { get; set; }
+
+        public string Availability { get; set; }
+
+        public string FriendlyName { get; set; }
+
+        public string DeviceName { get; set; }
+
+        public string DeviceNameFull { get; set; }
```
[the remark this answers is not in the document]
Comment (Averus-a) [reply src/Orchestra.Core/Windows/MonitorInfo.cs:45]: The DeviceNameFull contains actual windows name for screen, e.g.
DeviceName:  \\.\DISPLAY1
DeviceNameFull: \\.\DISPLAY1\Monitor0

File: src/Orchestra.Core/Windows/MonitorInfo.cs [hunk 5]
```diff
@@ -0,0 +1,452 @@
...
+
+        public Int32Rect MonitorArea { get; set; }
+
+        public Int32Rect WorkingArea { get; set; }
+
+        public DpiScale DpiScale { get; set; }
+
+        public int GetDpiAwareResolution()
+        {
+            throw new NotImplementedException();
```
[the remark this answers is not in the document]
Comment (Averus-a) [reply src/Orchestra.Core/Windows/MonitorInfo.cs]: Implemented in the last commit

File: src/Orchestra.Core/Windows/MonitorInfo.cs [hunk 6]
```diff
@@ -0,0 +1,452 @@
...
+                }
+            }
+
+            List<NativeMonitorInfoEx> nativeMonitorInfos = new List<NativeMonitorInfoEx>();
+
+            // Step 5: Enumerate available monitors
+            NativeMethods.EnumDisplayMonitors(IntPtr.Zero, IntPtr.Zero, delegate (IntPtr hMonitor, IntPtr hdcMonitor, ref Rect lprcMonitor, IntPtr dwData)
+            {
+                var mi = new NativeMonitorInfoEx();
+                var success = NativeMethods.GetMonitorInfo(hMonitor, mi);
```
[the remark this answers is not in the document]
Comment (Averus-a) [reply src/Orchestra.Core/Windows/MonitorInfo.cs]: It shouldn't as NativeMonitorInfoEx was a class and the marchaling was controlled directly on function with [In][Out] but i changed it to struct to be coherent with other native types (e.g. DisplayDevice), see last commit

File: src/Orchestra.Core/Windows/MonitorInfo.cs [hunk 7]
```diff
@@ -0,0 +1,507 @@
...
+            Y = absoluteDpiY / (double)BasicAbsoluteDpi;
+        }
+
+        public override string ToString()
+        {
+            return $"X:{X} Y:{Y}";
+        }
+    }
+
+    public class MonitorInfo
```
Comment (GeertvanHorrik) [inline src/Orchestra.Core/Windows/MonitorInfo.cs:33]: I think this class van be covered with some simple unit tests?

File: src/Orchestra.Core/Windows/MonitorInfo.cs [hunk 8]
```diff
@@ -0,0 +1,507 @@
...
+        public static MonitorInfo[] GetAllMonitors(bool throwErrorsForWrongAppManifest = true)
+        {
+            if (throwErrorsForWrongAppManifest)
+            {
+                // Step 1: check DPI awareness, must be PerMonitor
+                NativeMethods.GetProcessDpiAwareness(Process.GetCurrentProcess().Handle, out DpiAwareness awareness);
+
+                if (awareness != DpiAwareness.ProcessPerMonitor)
+                {
+                    throw new NotSupportedException("Application manifest is incorrect to retrieve reliable monitor info, see https://github.com/wildgums/orchestra/364 for more info");
```
Comment (GeertvanHorrik) [inline src/Orchestra.Core/Windows/MonitorInfo.cs:78]: throw Log.ErrorAndCrestrException<NotSupporredException>(...)

File: src/Orchestra.Core/Windows/MonitorInfo.cs [hunk 9]
```diff
@@ -0,0 +1,507 @@
...
+
+                if (!NativeMethods.EnumDisplayDevices(null, (uint)adapterIndex, ref adapter, 0))
+                {
+                    break;
+                }
+
+                videoAdapters.Add(adapter);
+            }
+
+            Dictionary<DisplayDevice, string> displayDevicesToAdapter = new Dictionary<DisplayDevice, string>();
```
Comment (GeertvanHorrik) [inline src/Orchestra.Core/Windows/MonitorInfo.cs:108]: Use var

File: src/Orchestra.Core/Windows/MonitorInfo.cs [hunk 10]
```diff
@@ -0,0 +1,507 @@
...
+                    if (!NativeMethods.EnumDisplayDevices(adapter.DeviceName, (uint)displayIndex, ref display, 1))
+                    {
+                        break;
+                    }
+
+                    displayDevicesToAdapter.Add(display, adapter.DeviceName);
+                }
+            }
+
+            Dictionary<NativeMonitorInfoEx, IntPtr> nativeMonitorInfos = new Dictionary<NativeMonitorInfoEx, IntPtr>();
```
Comment (GeertvanHorrik) [inline src/Orchestra.Core/Windows/MonitorInfo.cs:130]: Use var

File: src/Orchestra.Core/Windows/MonitorInfo.cs [hunk 11]
```diff
@@ -0,0 +1,507 @@
...
+            // Get screen from window
+            var windowInteropHelper = new WindowInteropHelper(window);
+            return GetMonitorFromWindowHandle(windowInteropHelper.Handle);
+        }
+
+        public static MonitorInfo GetMonitorFromWindowHandle(IntPtr handle)
+        {
+            if (handle == IntPtr.Zero)
+            {
+                throw new ArgumentException("Pointer has been initialized to zero", nameof(handle));
```
Comment (GeertvanHorrik) [inline src/Orchestra.Core/Windows/MonitorInfo.cs:204]: Throw Log.ErrorAndCreateException...

File: src/Orchestra.Core/Windows/MonitorInfo.cs [hunk 12]
```diff
@@ -0,0 +1,507 @@
...
+                DeviceNameFull = outputDevice.Value.DeviceName,
+                AdapterDeviceName = nativeInfo.GetDeviceName(),
+                DpiScale = dpiScale
+            };
+
+            return di;
+        }
+
+        private static DisplayDevice[] GetOutputDevicesForDevice(string adapterDeviceName)
+        {
```
Comment (GeertvanHorrik) [inline src/Orchestra.Core/Windows/MonitorInfo.cs:255]: Argument.IsNotNuklOrEmpty

File: src/Orchestra.Core/Windows/MonitorInfo.cs [hunk 13]
```diff
@@ -0,0 +1,507 @@
...
+                outputDevices.Add(display);
+            }
+
+            return outputDevices.ToArray();
+        }
+
+        internal static class NativeMethods
+        {
+            [DllImport("user32.dll", CharSet = CharSet.Unicode)]
+            public static extern bool GetMonitorInfo(IntPtr hmonitor, ref NativeMonitorInfoEx info);
```
Comment (GeertvanHorrik) [inline src/Orchestra.Core/Windows/MonitorInfo.cs:279]: hMonitor

File: src/Orchestra.Core/Windows/MonitorInfo.cs [hunk 14]
```diff
@@ -0,0 +1,507 @@
...
+            [DllImport("user32.dll")]
+            public static extern bool EnumDisplayMonitors(IntPtr hdc, IntPtr lprcClip, EnumMonitorsDelegate lpfnEnum, IntPtr dwData);
+
+            public delegate bool EnumMonitorsDelegate(IntPtr hMonitor, IntPtr hdcMonitor, ref Rect lprcMonitor, IntPtr dwData);
+
+            [DllImport("user32.dll")]
+            public static extern IntPtr MonitorFromWindow(IntPtr handle, int flags);
+
+            [DllImport("shcore.dll")]
+            public static extern bool GetDpiForMonitor([In] IntPtr hmonitor, [In] DpiType dpiType, [Out] out uint dpiX, [Out] out uint dpiY);
```
Comment (GeertvanHorrik) [inline src/Orchestra.Core/Windows/MonitorInfo.cs:293]: hMonitor

File: src/Orchestra.Core/Windows/MonitorInfo.cs [hunk 15]
```diff
@@ -0,0 +1,507 @@
...
+
+            [DllImport("shcore.dll", SetLastError = true)]
+            public static extern bool SetProcessDpiAwareness(DpiAwareness awareness);
+
+            [DllImport("shcore.dll", SetLastError = true)]
+            public static extern void GetProcessDpiAwareness(IntPtr hprocess, out DpiAwareness awareness);
```
Comment (GeertvanHorrik) [inline src/Orchestra.Core/Windows/MonitorInfo.cs:299]: hProcess

File: src/Orchestra.Core/Windows/MonitorInfo.cs [hunk 16]
```diff
@@ -201,7 +204,7 @@ public static MonitorInfo GetMonitorFromWindowHandle(IntPtr handle)
         {
             if (handle == IntPtr.Zero)
             {
-                throw new ArgumentException("Pointer has been initialized to zero", nameof(handle));
+                Log.ErrorAndCreateException((string errorMessage) => new ArgumentException(errorMessage, nameof(handle)), "Pointer has been initialized to zero");
```
Comment (GeertvanHorrik) [inline src/Orchestra.Core/Windows/MonitorInfo.cs:207]: Add throw

**Comment by GeertvanHorrik** [comment]
1 thing left and we are ready to merge :)

**Pull request merged**